### PR TITLE
add vk_memory_budget sample

### PR DESCRIPTION
**Pull request by gpx1000**
## Description

Add vk_memory_budget.  This sample demonstrates how to utilize already existing code (the Instancing sample) to determine how much memory is allocated and how much budget is remaining for the GPU device.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/README.md [hunk 1]
```diff
@@ -185,6 +185,11 @@ Push descriptors apply the push constants concept to descriptor sets. Instead of
 **Extension**: [```VK_EXT_debug_utils```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_EXT_debug_utils)<br/>
 Uses the debug utilities extension to name and group Vulkan objects (command buffers, images, etc.). This information makes debugging in tools like RenderDoc significantly easier.
 
+### [Memory Budget](./extensions/memory_budget)<br/>
+**Extension**: [```VK_EXT_memory_budget``](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.
```
Comment (SaschaWillems) [inline samples/README.md]: It looks like this line break is breaking the link. It doesn't display correct for me when displaying the markdown.

Comment (gpx1000) [reply samples/README.md]: good catch!

Thanks Sascha.

File: samples/extensions/memory_budget/memory_budget.cpp [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+	draw();
+	if (!paused || camera.updated)
+	{
+		update_uniform_buffer(delta_time);
+	}
+}
+
+void MemoryBudget::on_update_ui_overlay(vkb::Drawer &drawer)
+{
+	drawer.text("Total Memory Usage: %llu MB", device_memory_total_usage / 1000000);
```
Comment (asuessenbach) [inline samples/extensions/memory_budget/memory_budget.cpp]: I think, to have 1 MB, you need 1024 * 1024 B?

File: samples/extensions/memory_budget/memory_budget.cpp [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+	if (drawer.header("Memory Heap Details:"))
+	{
+		for (uint32_t i = 0; i < device_memory_heap_count; i++)
+		{
+			drawer.text("Memory Heap %lu: Usage: %llu MB | Budget: %llu MB", i, physical_device_memory_budget_properties.heapUsage[i] / 1000000, physical_device_memory_budget_properties.heapBudget[i] / 1000000);
+		}
+	}
+
+	if (drawer.header("Mesh Density Control"))
+		drawer.slider_int("Mesh Density", &mesh_density, 50, 8192);
```
Comment (asuessenbach) [inline samples/extensions/memory_budget/memory_budget.cpp]: mesh_density seems to be not used anywhere.
Did you want to control the size of the instance data?

**Review by asuessenbach — COMMENTED** [review]
I assume, your changes in check_copyright_headers.py are not related to the rest of the PR?
Besides that, having a static display of the memory consumption from some arbitrary point in time in the preparation step is nice to see, but not very informative. At least, determining the memory consumption should be moved to the end of the prepare step.
But I think, it would be much nicer, to have some slider or so, to allocate/free some memory and see corresponding changes in the UI.
Finally, it would also be nice to have some human readable version of the `propertyFlags` of each heap in the UI, telling you what kind of heap that is.

File: samples/extensions/memory_budget/memory_budget.cpp (hunk 2 above)
Comment (gpx1000) [reply samples/extensions/memory_budget/memory_budget.cpp]: Yep, good bug catch.  Will get this updated.

**Review by gary-sweet — COMMENTED** [review]
Builds ok for me and correctly reports that the extension isn't supported, but I can't test the actual sample.

File: samples/extensions/memory_budget/README.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+- distributed under the License is distributed on an "AS IS" BASIS,
+- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+- See the License for the specific language governing permissions and
+- limitations under the License.
+-
+-->
+
+# Memory Budget: Extended features
+
+This sample demonstrates how to incorporate Vulkan memory budget extension. Memory budget extension helps users to
```
Comment (asuessenbach) [inline samples/extensions/memory_budget/README.md]: ... incorporate **the** Vulkan ...

File: samples/extensions/memory_budget/README.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+sample the memory budget consumption on each heap of from the ```physical device```, and is able to tell
```
Comment (asuessenbach) [inline samples/extensions/memory_budget/README.md]: ... heap of from the ... ?

File: samples/extensions/memory_budget/README.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+In order to enable usage of memory budget extension features, an instance extension and a device extension were
+introduced in the constructor of the ```MemoryBudget``` class, where:
+
+````cpp
+add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+add_device_extension(VK_EXT_MEMORY_BUDGET_EXTENSION_NAME);
+````
+
+Memory properties related structure instances were defined and initialized in the header of ```MemoryBudget``` class,
```
Comment (asuessenbach) [inline samples/extensions/memory_budget/README.md]: ... of **the** ```MemoryBudget```...

File: samples/extensions/memory_budget/memory_budget.cpp [hunk 7]
```diff
@@ -0,0 +1,637 @@
...
+	// Initialize physical device memory budget properties structures variables
+	physical_device_memory_budget_properties.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MEMORY_BUDGET_PROPERTIES_EXT;
+	physical_device_memory_budget_properties.pNext = nullptr;
+	// Initialize physical device memory properties structure variables
+	device_memory_properties.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MEMORY_PROPERTIES_2;
+	device_memory_properties.pNext = &physical_device_memory_budget_properties;
+}
+
+const MemoryBudget::ConvertedMemory MemoryBudget::update_converted_memory(uint64_t input_memory)
+{
```
Comment (asuessenbach) [inline samples/extensions/memory_budget/memory_budget.cpp:157]: Maybe it's worth to have the data as float and print them with 2 digits after the decimal point?

File: samples/extensions/memory_budget/memory_budget.cpp [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+	switch (inputVkMemoryHeapFlag)
+	{
+		case VK_MEMORY_HEAP_DEVICE_LOCAL_BIT:
+			return "Device Local Bit";
+		case VK_MEMORY_HEAP_MULTI_INSTANCE_BIT:        // NOTICE THAT: enum value also represents "VK_MEMORY_HEAP_MULTI_INSTANCE_BIT_KHR"
+			return "Multiple Instance Bit";
+		default:
+			// In case that it does NOT correspond to device local memory
+			return "Non-local Heap Memory";
```
Comment (asuessenbach) [inline samples/extensions/memory_budget/memory_budget.cpp]: I assume, "Non-local Heap Memory" means something like "Host Local Heap Memory"?

File: samples/extensions/memory_budget/README.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+as ```VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MEMORY_PROPERTIES_2```, and its ```pNext``` chained in a pointer of the
+structure instance ```physical_device_memory_budget_properties```. Where, the ```sType```
+of ```physical_device_memory_budget_properties``` is defined
+as ```VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MEMORY_BUDGET_PROPERTIES_EXT``` and its ```pNext``` is defined as
+a ```nullptr```.
+
+Each of the mentioned extension structure instances were defined in the header and initialized in
+function ```initialize_device_memory_properties()```.
+
+## UI Overlay: Memory properties
```
Comment (JoseEmilio-ARM) [inline samples/extensions/memory_budget/README.md:69]: should the .png screenshot be embedded here?

**Review by asuessenbach — APPROVED** [review]

**Review by SaschaWillems — APPROVED** [review]

**Pull request merged**